### ISSUE TITLE
Yet another issue with ChronicleHash

### ISSUE THREAD

**Issue by rkouznetsov** [Unidata/tds#317]
```
Manifest-Version: 1.0
Implementation-Title: THREDDS Data Server (TDS)
Implementation-Version: 5.5-SNAPSHOT
Implementation-Vendor-Id: edu.ucar
Implementation-Vendor: UCAR/Unidata
Implementation-URL: https://www.unidata.ucar.edu/software/tds/
Created-By: Gradle 6.9.1
Build-Jdk: 11.0.12
Built-By: ubuntu
Built-On: 2022-10-21T06:29:23+0000
```
Fails on FMRC initialisation. Could be related to https://github.com/Unidata/tds/issues/275, but the callstack is a bit longer:

```
[2022-11-17T08:00:03.736+0200] ERROR ucar.nc2.ft.fmrc.Fmrc: makeFmrcInv
com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalStateException: ChronicleMap{name=GridDatasetInv, file=/usr/share/tomcat/apache-tomcat-8.5.75/${tds.content.root.path}/thredds/cache/collection/GridDatasetInv.dat, identityHashCode=1112601959}: Attempt to allocate #321 extra segment tier, 320 is maximum.
Possible reasons include:
 - you have forgotten to configure (or configured wrong) builder.entries() number
 - same regarding other sizing Chronicle Hash configurations, most likely maxBloatFactor(), averageKeySize(), or averageValueSize()
 - keys, inserted into the ChronicleHash, are distributed suspiciously bad. This might be a DOS attack
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051) ~[guava-30.1-jre.jar:?]
        at com.google.common.cache.LocalCache.get(LocalCache.java:3951) ~[guava-30.1-jre.jar:?]
        at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4848) ~[guava-30.1-jre.jar:?]
        at ucar.nc2.ft.fmrc.GridDatasetInv.open(GridDatasetInv.java:72) ~[cdm-core-5.5.4-SNAPSHOT.jar:5.5.4-SNAPSHOT]
        at ucar.nc2.ft.fmrc.Fmrc.makeFmrcInv(Fmrc.java:287) [cdm-core-5.5.4-SNAPSHOT.jar:5.5.4-SNAPSHOT]
        at ucar.nc2.ft.fmrc.Fmrc.update(Fmrc.java:222) [cdm-core-5.5.4-SNAPSHOT.jar:5.5.4-SNAPSHOT]
        at thredds.featurecollection.InvDatasetFcFmrc.updateCollection(InvDatasetFcFmrc.java:140) [classes/:5.5-SNAPSHOT]
        at thredds.featurecollection.InvDatasetFeatureCollection.checkState(InvDatasetFeatureCollection.java:280) [classes/:5.5-SNAPSHOT]
        at thredds.featurecollection.InvDatasetFcFmrc.getGridDataset(InvDatasetFcFmrc.java:460) [classes/:5.5-SNAPSHOT]
        at thredds.featurecollection.InvDatasetFeatureCollection.getNetcdfDataset(InvDatasetFeatureCollection.java:586) [classes/:5.5-SNAPSHOT]
        at thredds.featurecollection.InvDatasetFcFmrc.getGridCoverage(InvDatasetFcFmrc.java:443) [classes/:5.5-SNAPSHOT]
        at thredds.core.DatasetManager.openCoverageDataset(DatasetManager.java:416) [classes/:5.5-SNAPSHOT]
        at thredds.core.TdsRequestedDataset.openAsCoverageDataset(TdsRequestedDataset.java:142) [classes/:5.5-SNAPSHOT]
        at thredds.core.TdsRequestedDataset.getCoverageCollection(TdsRequestedDataset.java:72) [classes/:5.5-SNAPSHOT]
        at thredds.server.ncss.controller.NcssGridController.handleRequest(NcssGridController.java:84) [classes/:5.5-SNAPSHOT]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
        at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205) [spring-web-5.3.20.jar:5.3.20]
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150) [spring-web-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117) [spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895) [spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808) [spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) [spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1067) [spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963) [spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) [spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898) [spring-webmvc-5.3.20.jar:5.3.20]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:655) [servlet-api.jar:?]
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) [spring-webmvc-5.3.20.jar:5.3.20]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:764) [servlet-api.jar:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) [catalina.jar:8.5.75]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.75]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) [tomcat-websocket.jar:8.5.75]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.75]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.75]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:327) [spring-security-web-5.6.5.jar:5.6.5]
        at thredds.servlet.filter.RequestBracketingLogMessageFilter.doFilter(RequestBracketingLogMessageFilter.java:50) [classes/:5.5-SNAPSHOT]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336) [spring-security-web-5.6.5.jar:5.6.5]
        at thredds.servlet.filter.RequestQueryFilter.doFilter(RequestQueryFilter.java:90) [classes/:5.5-SNAPSHOT]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336) [spring-security-web-5.6.5.jar:5.6.5]
        at thredds.servlet.filter.HttpHeadFilter.doFilter(HttpHeadFilter.java:47) [classes/:5.5-SNAPSHOT]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336) [spring-security-web-5.6.5.jar:5.6.5]
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:211) [spring-security-web-5.6.5.jar:5.6.5]
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:183) [spring-security-web-5.6.5.jar:5.6.5]
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:354) [spring-web-5.3.20.jar:5.3.20]
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:267) [spring-web-5.3.20.jar:5.3.20]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.75]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.75]
        at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71) [log4j-web-2.17.1.jar:2.17.1]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.75]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.75]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:196) [catalina.jar:8.5.75]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) [catalina.jar:8.5.75]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542) [catalina.jar:8.5.75]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135) [catalina.jar:8.5.75]
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:698) [catalina.jar:8.5.75]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81) [catalina.jar:8.5.75]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) [catalina.jar:8.5.75]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:366) [catalina.jar:8.5.75]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:639) [tomcat-coyote.jar:8.5.75]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) [tomcat-coyote.jar:8.5.75]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:847) [tomcat-coyote.jar:8.5.75]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1680) [tomcat-coyote.jar:8.5.75]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-coyote.jar:8.5.75]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) [tomcat-util.jar:8.5.75]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) [tomcat-util.jar:8.5.75]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-util.jar:8.5.75]
        at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.IllegalStateException: ChronicleMap{name=GridDatasetInv, file=/usr/share/tomcat/apache-tomcat-8.5.75/${tds.content.root.path}/thredds/cache/collection/GridDatasetInv.dat, identityHashCode=1112601959}: Attempt to allocate #321 extra segment tier, 320 is maximum.
Possible reasons include:
 - you have forgotten to configure (or configured wrong) builder.entries() number
 - same regarding other sizing Chronicle Hash configurations, most likely maxBloatFactor(), averageKeySize(), or averageValueSize()
 - keys, inserted into the ChronicleHash, are distributed suspiciously bad. This might be a DOS attack
        at net.openhft.chronicle.hash.impl.VanillaChronicleHash.allocateTier(VanillaChronicleHash.java:822) ~[chronicle-map-3.21ea6.jar:3.21ea6]
        at net.openhft.chronicle.map.impl.CompiledMapQueryContext.nextTier(CompiledMapQueryContext.java:3131) ~[chronicle-map-3.21ea6.jar:3.21ea6]
        at net.openhft.chronicle.map.impl.CompiledMapQueryContext.alloc(CompiledMapQueryContext.java:3492) ~[chronicle-map-3.21ea6.jar:3.21ea6]
        at net.openhft.chronicle.map.impl.CompiledMapQueryContext.initEntryAndKey(CompiledMapQueryContext.java:3510) ~[chronicle-map-3.21ea6.jar:3.21ea6]
        at net.openhft.chronicle.map.impl.CompiledMapQueryContext.putEntry(CompiledMapQueryContext.java:4003) ~[chronicle-map-3.21ea6.jar:3.21ea6]
        at net.openhft.chronicle.map.impl.CompiledMapQueryContext.doInsert(CompiledMapQueryContext.java:4192) ~[chronicle-map-3.21ea6.jar:3.21ea6]
        at net.openhft.chronicle.map.MapEntryOperations.insert(MapEntryOperations.java:153) ~[chronicle-map-3.21ea6.jar:3.21ea6]
        at net.openhft.chronicle.map.impl.CompiledMapQueryContext.insert(CompiledMapQueryContext.java:4115) ~[chronicle-map-3.21ea6.jar:3.21ea6]
        at net.openhft.chronicle.map.MapMethods.put(MapMethods.java:88) ~[chronicle-map-3.21ea6.jar:3.21ea6]
        at net.openhft.chronicle.map.VanillaChronicleMap.put(VanillaChronicleMap.java:856) ~[chronicle-map-3.21ea6.jar:3.21ea6]
        at thredds.featurecollection.cache.GridInventoryCacheChronicle.put(GridInventoryCacheChronicle.java:104) ~[classes/:5.5-SNAPSHOT]
        at ucar.nc2.ft.fmrc.GridDatasetInv$GenerateInv.call(GridDatasetInv.java:122) ~[cdm-core-5.5.4-SNAPSHOT.jar:5.5.4-SNAPSHOT]
        at ucar.nc2.ft.fmrc.GridDatasetInv$GenerateInv.call(GridDatasetInv.java:78) ~[cdm-core-5.5.4-SNAPSHOT.jar:5.5.4-SNAPSHOT]
        at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4853) ~[guava-30.1-jre.jar:?]
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529) ~[guava-30.1-jre.jar:?]
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278) ~[guava-30.1-jre.jar:?]
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155) ~[guava-30.1-jre.jar:?]
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045) ~[guava-30.1-jre.jar:?]
        ... 69 more
```

As a result all requests result in Error 500:
```
CH_m22,heatsum_POLLEN_BIRCH_m22,poll_left_POLLEN_BIRCH_m22,poll_tot_m2_POLLEN_BIRCH_m22,pollen_corr_POLLEN_BIRCH_m22,cnc_POLLEN_BIRCH_m22,Poll_Rdy2fly_POLLEN_GRASS_m32,poll_left_POLLEN_GRASS_m32,poll_tot_m2_POLLEN_GRASS_m32,pollen_corr_POLLEN_GRASS_m32,cnc_POLLEN_GRASS_m32,Poll_Rdy2fly_POLLEN_MUGWORT_m18,poll_left_POLLEN_MUGWORT_m18,cnc_POLLEN_MUGWORT_m18,Poll_Rdy2fly_POLLEN_MUGW1_m18,poll_left_POLLEN_MUGW1_m18,cnc_POLLEN_MUGW1_m18,Poll_Rdy2fly_POLLEN_MUGW2_m18,poll_left_POLLEN_MUGW2_m18,cnc_POLLEN_MUGW2_m18,Poll_Rdy2fly_POLLEN_MUGW3_m18,poll_left_POLLEN_MUGW3_m18,cnc_POLLEN_MUGW3_m18,Poll_Rdy2fly_POLLEN_MUGW4_m18,poll_left_POLLEN_MUGW4_m18,cnc_POLLEN_MUGW4_m18,Poll_Rdy2fly_POLLEN_MUGW5_m18,poll_left_POLLEN_MUGW5_m18,cnc_POLLEN_MUGW5_m18,Poll_Rdy2fly_POLLEN_OLIVE_m28,heatsum_POLLEN_OLIVE_m28,poll_left_POLLEN_OLIVE_m28,poll_tot_m2_POLLEN_OLIVE_m28,pollen_corr_POLLEN_OLIVE_m28,cnc_POLLEN_OLIVE_m28,Poll_Rdy2fly_POLLEN_RAGWEED_m18,heatsum_POLLEN_RAGWEED_m18,poll_left_POLLEN_RAGWEED_m18,poll_tot_m2_POLLEN_RAGWEED_m18,pollen_corr_POLLEN_RAGWEED_m18,cnc_POLLEN_RAGWEED_m18&spatial=bb&north=56&west=4&east=16&south=46&temporal=range&time_start=2022-11-17T20:00:00Z&time_end=2022-11-17T20:00:00Z&accept=netcdf3&email=test-download-bavaria HTTP/1.1"
2022-11-17T09:37:57.827 +0200 [  40460700][    2725] ERROR - thredds.server.TdsErrorHandling - uncaught exception
java.lang.NullPointerException: null
        at ucar.nc2.ft.fmrc.FmrcDataset.getRunTimeDataset(FmrcDataset.java:196) ~[cdm-core-5.5.4-SNAPSHOT.jar:5.5.4-SNAPSHOT]
        at ucar.nc2.ft.fmrc.Fmrc.getRunTimeDataset(Fmrc.java:189) ~[cdm-core-5.5.4-SNAPSHOT.jar:5.5.4-SNAPSHOT]
        at thredds.featurecollection.InvDatasetFcFmrc.getGridDataset(InvDatasetFcFmrc.java:507) ~[classes/:5.5-SNAPSHOT]
        at thredds.featurecollection.InvDatasetFeatureCollection.getNetcdfDataset(InvDatasetFeatureCollection.java:586) ~[classes/:5.5-SNAPSHOT]
        at thredds.featurecollection.InvDatasetFcFmrc.getGridCoverage(InvDatasetFcFmrc.java:443) ~[classes/:5.5-SNAPSHOT]
        at thredds.core.DatasetManager.openCoverageDataset(DatasetManager.java:416) ~[classes/:5.5-SNAPSHOT]
        at thredds.core.TdsRequestedDataset.openAsCoverageDataset(TdsRequestedDataset.java:142) ~[classes/:5.5-SNAPSHOT]
        at thredds.core.TdsRequestedDataset.getCoverageCollection(TdsRequestedDataset.java:72) ~[classes/:5.5-SNAPSHOT]
        at thredds.server.ncss.controller.NcssGridController.handleRequest(NcssGridController.java:84) ~[classes/:5.5-SNAPSHOT]
        at jdk.internal.reflect.GeneratedMethodAccessor94.invoke(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
        at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205) ~[spring-web-5.3.20.jar:5.3.20]
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150) ~[spring-web-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117) ~[spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895) ~[spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808) ~[spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1067) [spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963) [spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) [spring-webmvc-5.3.20.jar:5.3.20]
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898) [spring-webmvc-5.3.20.jar:5.3.20]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:655) [servlet-api.jar:?]
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) [spring-webmvc-5.3.20.jar:5.3.20]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:764) [servlet-api.jar:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) [catalina.jar:8.5.75]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.75]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) [tomcat-websocket.jar:8.5.75]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.75]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.75]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:327) [spring-security-web-5.6.5.jar:5.6.5]
        at thredds.servlet.filter.RequestBracketingLogMessageFilter.doFilter(RequestBracketingLogMessageFilter.java:50) [classes/:5.5-SNAPSHOT]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336) [spring-security-web-5.6.5.jar:5.6.5]
        at thredds.servlet.filter.RequestQueryFilter.doFilter(RequestQueryFilter.java:90) [classes/:5.5-SNAPSHOT]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336) [spring-security-web-5.6.5.jar:5.6.5]
        at thredds.servlet.filter.HttpHeadFilter.doFilter(HttpHeadFilter.java:47) [classes/:5.5-SNAPSHOT]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336) [spring-security-web-5.6.5.jar:5.6.5]
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:211) [spring-security-web-5.6.5.jar:5.6.5]
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:183) [spring-security-web-5.6.5.jar:5.6.5]
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:354) [spring-web-5.3.20.jar:5.3.20]
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:267) [spring-web-5.3.20.jar:5.3.20]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.75]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.75]
        at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71) [log4j-web-2.17.1.jar:2.17.1]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.75]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.75]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:196) [catalina.jar:8.5.75]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) [catalina.jar:8.5.75]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542) [catalina.jar:8.5.75]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135) [catalina.jar:8.5.75]
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:698) [catalina.jar:8.5.75]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81) [catalina.jar:8.5.75]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) [catalina.jar:8.5.75]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:366) [catalina.jar:8.5.75]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:639) [tomcat-coyote.jar:8.5.75]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) [tomcat-coyote.jar:8.5.75]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:847) [tomcat-coyote.jar:8.5.75]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1680) [tomcat-coyote.jar:8.5.75]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-coyote.jar:8.5.75]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) [tomcat-util.jar:8.5.75]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) [tomcat-util.jar:8.5.75]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-util.jar:8.5.75]
        at java.lang.Thread.run(Thread.java:829) [?:?]
2022-11-17T09:37:57.853 +0200 [  40460726][    2725] INFO  - threddsServlet - Request Completed - 500 - -1 - 27
```

I have further increased maxEntries from 10k to 50k:
```
 <FeatureCollection>
  <dir>${tds.content.root.path}/thredds/cache/collection/</dir>
  <maxEntries>50000</maxEntries>
  <maxBloatFactor>10</maxBloatFactor>
</FeatureCollection>
```
 in `threddsConfig.xml` but it did not have any effect. Neither did manual clean of the cache directories and restarting tomcat.

Currently the data directory contains about 18k files, half of them are symlinks listed in thredds catalogs.

Thank you!

**Comment by rkouznetsov** [comment]
Might be related to https://github.com/OpenHFT/Chronicle-Map/issues/192

**Comment by haileyajohnson** [comment]
This issue was solve by #232 
See [here](https://docs.unidata.ucar.edu/tds/current/userguide/tds_config_ref.html#featurecollection-cache) for how to configure your cache size.

**Comment by haileyajohnson** [comment]
Sorry I missed the last part that the config wasn't working out!

**Comment by tdrwenski** [comment]
Can you give me an idea or example of how long your file paths are for your FMRC files? The cache uses the file path as the cache key, so maybe a much longer than expected file path could be causing issues with the cache allocation.

**Comment by rkouznetsov** [comment]
The longest file path used in catalog is 104 bytes.

**Comment by rkouznetsov** [comment]
I do not know if it is of any relevance.. I have just spotted that I have a _file_ with a name
`'/usr/share/tomcat/apache-tomcat-8.5.75/${tds.content.root.path}/thredds/cache/collection/GridDatasetInv.dat'`  with a timestamp corresponding to the time of last tomcat restart.
```
$ ls -la /usr/share/tomcat/apache-tomcat-8.5.75/\$\{tds.content.root.path\}/thredds/cache/collection/GridDatasetInv.dat 
-rw-r-----. 1 tomcat tomcat 711647232 Nov 18 18:14 '/usr/share/tomcat/apache-tomcat-8.5.75/${tds.content.root.path}/thredds/cache/collection/GridDatasetInv.dat'
```
Looks like tomcat/tds forgot to substitute some variable...

**Comment by tdrwenski** [comment]
Thanks for the extra info, that is indeed weird-- we will have a look!

**Comment by rkouznetsov** [comment]
After removing that file and restarting tomcat the server started responding. The file re-appeared again though.

**Comment by tdrwenski** [comment]
what if you remove the line `<dir>${tds.content.root.path}/thredds/cache/collection/</dir>` from your FeatureCollection in the threddsConfig.xml? That should be the same as the default, but possibly the variable is failing to be substituted there.

**Comment by rkouznetsov** [comment]
I see now. The path in https://github.com/Unidata/tds/issues/275#issuecomment-1248327796 was literally reproduced. Now I have put there a path without a reference to `${tds.content.root.path}` variable, and it seems to work. 

So once the maxEntries limit was reached the server got stuck. Reviving it required manual deletion of cache.  `maxBloatFactor` is ignored, or attempt to use it puts cache to an invalid state...

**Comment by tdrwenski** [comment]
Okay, I will update our docs to not have
`<dir>${tds.content.root.path}/thredds/cache/collection/</dir>`
in our example, as that is confusing and can lead to a hidden cache that is then harder to manually delete.

I am not sure why the `maxBloatFactor` > 1 is not helping increase the cache. With a high enough `maxEntries` is it now working for you?

**Comment by rkouznetsov** [comment]
Thank you! So the variable is not supposed to be expanded in `threddsConfig.xml`, right? Looks like then the line with cache path is not needed at all if one uses the default path. 

The TDS-5.5 server has not yet in production due to several other open issues, bit currently it at least can serve some requests, so I re-enabled test operations for it. TDS4.6 is still a working horse for us.

With the current issue, besides a bit confusing example, there is actually two problems:
- apparently non-working `maxBloatFactor` 
- cache entering an invalid state, from which the server cannot recover without manual removal of the cache files

I wonder if it would be possible to disable caching with CronicleHash completely, or at least enable extensive logging for it... I believe, it might also help hunting down https://github.com/Unidata/tds/issues/293...

**Comment by tdrwenski** [comment]
It looks like the `threddsConfig.xml` does not currently expand any variables. I have updated the docs to not have examples with variables.

Sorry to hear that this issue and your other open issue are preventing you from using TDS 5 in production. We definitely want to get these issues resolved soon.

For this issue, thanks for the extra details! I will try to test the `maxBloatFactor` and see if we can add something to better manage or disable this cache like you suggest.

**Comment by rkouznetsov** [comment]
Thank you! No worries. Please take your time to resolve the issues thoroughly.  

Unlike some years ago, the issues are now public, so I can use the list of open issues to convince our IT folks that we really need to keep both servers up and running for a while. Our routines to  manage them both seem to work fine so far.

I hope, our tests will also be beneficial for TDS development.

**Comment by tdrwenski** [comment]
I tested with one of your data files `SILAM-AQ-glob06_v5_8_2022090900_002.nc4` and I found that it has a lot of grids so it takes up 18 kB in the cache whereas we assume the average value size is 4 kB. I think this is why even with the correct `maxEntries` and a high `maxBloatFactor` the cache isn't properly increasing in size. So we need to either add another configuration element or do something smarter when initializing the cache.

Is this a reasonable file to test with? And if so, would you mind if I added it to our test data?

**Comment by rkouznetsov** [comment]
Thank you! The file has one grid and few variables that do not have a gridd (in geographical sense)...  However `cdo griddes` reports three grids for he file:
```
#
# gridID 1
#
gridtype  = lonlat
gridsize  = 178200
datatype  = float
xsize     = 600
ysize     = 297
xname     = lon
xlongname = "longitude"
xunits    = "degrees_east"
yname     = lat
ylongname = "latitude"
yunits    = "degrees_north"
xfirst    = -179.4
xinc      = 0.6
yfirst    = -88.8
yinc      = 0.6
scanningMode = 64
#
# gridID 2
#
gridtype  = generic
gridsize  = 1
scanningMode = 64
#
# gridID 3
#
gridtype  = generic
gridsize  = 29
xsize     = 29
scanningMode = 64
```
The latter two are just dimensions for supplementary variables, such as layer-bound hybrid coefficients (aka half-levels) and their increments . 

Please feel free to use the file as a test case, but I would first reduce its size by removing  all 2D variables and all-but one one truly-3D variable, and, probably cutting grid to some 8x8 cells to make the file nice and small.

**Comment by tdrwenski** [comment]
We have added an extra thredds config parameter in the latest snapshot that should hopefully fix your issue. The snapshot can be downloaded [here](https://downloads.unidata.ucar.edu/tds/5.5/thredds-5.5-SNAPSHOT.war). The documentation should be updated soon as well.

The new parameter is called `averageValueSize` and can have values `small` (default), `medium`, and `large`. I believe that setting this to `medium` should be enough to fix your issue, for instance like:
```
<FeatureCollection>
  <maxEntries>50000</maxEntries>
  <maxBloatFactor>10</maxBloatFactor>
  <averageValueSize>medium</averageValueSize>
 </FeatureCollection>
 ```
 
 In addition, we added a bit of information about the FMRC cache to the [admin debug page](https://docs.unidata.ucar.edu/tds/5.4/userguide/remote_management_ref.html#tds-remote-debugging) under "show caches" (thredds/admin/debug?Caches/showCaches). I hope that helps with future debugging.

Let us know if you still have issues!

**Comment by rkouznetsov** [comment]
Thank you!  We have updated the binary and added the magic you recommend. 

I could not really tell if that changed much in the server behavior. At least no uncaught exceptions in ChronicleHash so far (https://github.com/Unidata/tds/issues/325 is still there).  cache.log besides start-stop messages shows
```
2-12-05T15:06:08.234+0200] Started a DiskCache2 scour task on /content/thredds/cache/agg every 1440 minutes for files older than 129600 minutes
2-12-05T15:06:08.235+0200] Started a DiskCache2 scour task on /content/thredds/cache/aggNew/ every 1440 minutes for files older than 129600 minutes
2-12-05T15:06:08.284+0200] Started a DiskCache2 scour task on /content/thredds/cache/ncss every 15 minutes for files older than 30 minutes
2-12-05T15:06:09.816+0200] Started a DiskCache2 scour task on /content/thredds/cache/wcs every 10 minutes for files older than 5 minutes
2-12-06T04:01:55.253+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 151
2-12-06T04:01:55.435+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 152
2-12-06T04:01:55.662+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 153
2-12-06T04:01:55.878+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 154
2-12-06T04:01:56.033+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 155
2-12-06T04:01:56.184+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 156
2-12-06T04:01:56.342+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 157
2-12-06T04:01:56.925+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 158
2-12-06T04:01:57.106+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 159
2-12-06T04:01:57.263+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 160
2-12-06T04:01:57.451+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 161
2-12-06T04:01:57.679+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 162
2-12-06T04:01:57.918+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 163
2-12-06T04:01:58.096+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 164
2-12-06T04:01:58.275+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 165
2-12-06T04:01:58.423+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 166
2-12-06T04:01:58.614+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 167
2-12-06T04:01:58.819+0200] FileCache NetcdfFileCache (deprecated) cleanup couldnt remove enough to keep under the maximum= 150 due to locked files; currently at = 168
```
I wonder if it is related...

As a general question:  is the exceeding of the requested cache size considered as a legitimate reason for the server to crash?

**Comment by tdrwenski** [comment]
Glad you aren't seeing any ChronicleMap errors so far! You say you can't tell if the behavior changed much, does that mean you are still having issues with the FMRC datasets?

For the messages you are seeing in the cache.log-- That is about the NetcdfFileCache which does not use ChronicleMap like the FMRC cache, so I don't think its related. I don't think that should cause crashes, though it may cause sub-optimal performance as it can't put the files it needs to into the cache. It sounds like some cache files can never be removed because they are locked. Files should only be locked while in use, so a bunch of locked ones may indicate a problem.

I recently found this issue or a very similar one and am [working on a fix](https://github.com/Unidata/netcdf-java/pull/1120). Maybe once that is merged (I hope later this week) you could check if those messages are resolved and if not, feel free to open a new issue about it. You can also check with the admin debug service what files are in the cache and if they are locked or unlocked. I hope that helps!

**Comment by rkouznetsov** [comment]
Thank you! 

> does that mean you are still having issues with the FMRC datasets?
Not with ChronicleMap so far :) I just meant to say that the issue now occurs rarely enough to reliably tell that it is gone.
Besides that I have not tested what happens if the requested ChronicleMap cache size is exceeded.

I'll be happy to test if your bug fix cleans that.

Just a small question. I have not managed to setup an access to the admin debug page. I have made a user with a password with tomcat-users.xml, but I keep getting 403 error when trying to access the page. Anyway, I would prefer to have all the debug info in logs. Does that page has anything beyond the stuff that comes to cache.log, and if not, how to make that info to logs?

**Comment by tdrwenski** [comment]
Okay, great thanks! I think I will close this issue then. I also appreciate your help testing that NetcdfFileCache issue!

Regarding the admin debug service-- I do think it has some info not in the logs, (like what is currently in the cache) but it is certainly not necessary to use it! Let us know if you want assistance with setting it up.

**Comment by rkouznetsov** [comment]
Thank you!

**Comment by tdrwenski** [comment]
@rkouznetsov the fix that I mentioned for the NetcdfFileCache is in the latest snapshot if you want to test if that fixes your issue.
